docs/blog/2019-03-18-releasing-new-schema-customization/index.md

```diff
@@ -0,0 +1,127 @@
+
+```js:title=gatsby-node.js
+exports.sourceNodes = ({ actions }) => {
+  const { createTypes } = actions
+  const typeDefs = `
+    type AuthorJson implements Node {
+      name: String
+      birthday: Date
+    }
+  `
+  createTypes(typeDefs)
+}
+```
+
+After adding this to your gatsby-node, `AuthorJson` type will always have fields name and birthday, regardless of the automatically inferred data shape. The rest of the fields will still be inferred normally, allowing you to still enjoy the benefits of Gatsby schema inference.
+
+## `createResolvers`
+
+`createResolvers` allows doing additional customization after all schema processing has been finished. Thus it can be used to add fields to any types, including root types like `Query` and types from third party schemas.
+
+```js:title=gatsby-node.js
+createResolvers({
+  Query: {
+    allAuthorFullNames: {
+      type: `[String!]!`,
+      resolve(source, args, context, info) {
+        const authors = context.nodeModel.getAllNodes({
+          type: `AuthorJson`,
+        })
+        return authors.map(author => author.name)
+      },
+    },
+  },
+})
+```
+
+## The Type Builder API
+
+While `createTypes` accepts `graphql-js` types along with SDL string, we've also added an option to use `graphql-js` types so that user could create types with resolvers. However, `graphql-js` is pretty verbose and it's hard to refer to types that don't yet exist or don't exist in a current scope. Therefore, we decided to add another programmatic API, that combines brevity of SDL with flexibility of `graphql-js`.
+
+We can this API _Type Builder API_. It is available in the `schema` field of the arguments object passed to [Gatsby Node APIs](/docs/node-apis/).
+
+```js:title=gatsby-node.js
+exports.sourceNodes = ({ actions, schema }) => {
+  const { createTypes } = actions
+  createTypes([
+    schema.buildObjectType({
+      name: `CommentJson`,
+      fields: {
+        text: `String!`,
+        blog: {
+          type: `BlogJson`,
+          resolve(parent, args, context) {
+            return context.nodeModel.getNodeById({
+              id: parent.author,
+              type: `BlogJson`,
+            })
+          },
+        },
+        author: {
+          type: `AuthorJson`,
+          resolve(parent, args, context) {
+            return context.nodeModel.getNodeById({
+              id: parent.author,
+              type: `AuthorJson`,
+            })
+          },
+        },
+      },
+      interfaces: [`Node`],
+    }),
+  ])
+}
+```
+
+# Potential for Breaking Changes
+
+We have tried to avoid any breaking changes in this refactor of the underlying GraphQL layer, testing it in notable Gatsby sites and ensuring all tests were passing. However, there are areas where we needed to introduce more stable naming, and in these instances it _could_ be possible that a breaking change was introduced if you were using this undocumented API.
```

Comment:
```suggestion
We have tried to avoid any breaking changes in this refactor of the underlying GraphQL layer, testing it in notable Gatsby sites and ensuring all tests were passing. However, there are areas where we needed to introduce more stable naming, and in these instances it _could_ be possible that a breaking change was introduced if you were relying on this undocumented API.
```